internal/v1/oscap.go

```diff
@@ -85,35 +59,117 @@ func OscapProfiles(distribution Distributions) (DistributionProfileResponse, err
+	// Convert the custom data structure into a `Customizations` object.
+	// This will be easier to handle in IB's API later on
+	customizations := Customizations{}
+
+	var fs []Filesystem
+	for _, bpFileSystem := range bp.Customizations.Filesystem {
+		fs = append(fs, Filesystem{
+			MinSize:    bpFileSystem.Size,
+			Mountpoint: bpFileSystem.Mountpoint,
+		})
+	}
+	if len(fs) > 0 {
+		customizations.Filesystem = &fs
 	}
-	defer jsonFile.Close()
-	bytes, err := io.ReadAll(jsonFile)
+
+	var packages []string
+	for _, bpPackage := range bp.Packages {
+		packages = append(packages, bpPackage.Name)
+	}
+
+	var kernel *Kernel
+	if k := bp.Customizations.Kernel; k != nil {
+		kernel = &Kernel{}
+		if k.Name != nil {
+			kernel.Name = k.Name
+		}
+		if k.Append != nil {
+			kernel.Append = k.Append
+		}
+	}
+	if kernel != nil {
+		customizations.Kernel = kernel
+	}
+
+	var services *Services
+	if s := bp.Customizations.Services; s != nil {
+		services = &Services{}
+		if s.Enabled != nil {
+			firewalldPkg := "firewalld"
+			if slices.Contains(*s.Enabled, firewalldPkg) && !slices.Contains(packages, firewalldPkg) {
+				packages = append(packages, firewalldPkg)
+			}
+			services.Enabled = s.Enabled
+		}
+		var maskedAndDisabled []string
+		if s.Disabled != nil {
+			maskedAndDisabled = append(maskedAndDisabled, *s.Disabled...)
+		}
+		if s.Masked != nil {
+			maskedAndDisabled = append(maskedAndDisabled, *s.Masked...)
+		}
+		// we need to collect both disabled and masked services and
+		// assign them to the masked customization, since disabled services
+		// that aren't installed on the image will break the image build.
+		if maskedAndDisabled != nil {
+			services.Masked = &maskedAndDisabled
+		}
+	}
+	if services != nil {
+		customizations.Services = services
+	}
+
+	if len(packages) > 0 {
+		customizations.Packages = &packages
+	}
+
+	profileDescription := bp.Description
+	if description != "" {
+		profileDescription = description
+	}
+
+	var openscap OpenSCAP
+	err := openscap.FromOpenSCAPProfile(OpenSCAPProfile{
+		ProfileId:          profile,
+		ProfileName:        &bp.Description, // annoyingly the Profile name is saved to the blueprint description
+		ProfileDescription: &profileDescription,
+	})
 	if err != nil {
 		return nil, err
 	}
-	// The customizations json file already contains a valid Customizations object to be returned as is.
-	var customizations Customizations
-	err = json.Unmarshal(bytes, &customizations)
+	customizations.Openscap = &openscap
+
+	return &customizations, nil
+}
+
+func processRequest(profile string, datastream string, tailoring *string) (*Customizations, error) {
+	// the generated blueprint doesn't contain the profile
+	// description, so we have to run the oscap tool to get
+	// this information
+	description := oscap.GetProfileDescription(profile, datastream)
+
+	var file *os.File
+	if tailoring != nil {
+		file, err := oscap.GenXMLTailoringFile(tailoring, datastream)
```

Comment:
Ditto, the file is never closed thus kernel will keep the file handle until restart.

Reply:
Whoops, thanks for catching this